### PR TITLE
Improve import time

### LINKED ISSUE #2154: Import time is too long
Importing something from DESC takes too much time. Specifically the 
```python
import desc.compute
```

Script to see individual contributions:
```
python -X importtime -c "import desc.compute"
```
This takes more than 5.5 seconds, and 3.5 seconds of it is just `_build_data_index`. `desc.integrals` is also taking around 2 seconds.

[import-time.txt](https://github.com/user-attachments/files/26609765/import-time.txt)

### PR DESCRIPTION
**Pull request by YigitElma**
- Replaces the recursive tree construction by a depth-first search algorithm in `desc.compute`
- Removes the `default_quad` computation at import time.

Importing `desc.compute` takes a very long time, and on my system, only `_build_data_index` itself takes 3.5 seconds. This change makes it run in 0.15 seconds!

The previous function was slow because of recursion, but it also had an extreme amount of redundant calls. For example, `get_params`, `get_profiles`, and `get_derivs` call `get_deps` again internally, and since the result of the first call is not stored yet, it rebuilds the whole tree again, almost 8 more times.

New `_build_data_index` was originally written by Claude Code (Opus 4.6), but I wrote most of the comments and I understand the algorithm behind.

Resolves #2154

### CODE REVIEW

File: desc/integrals/bounce_integral.py [hunk 1]
```diff
@@ -301,7 +295,16 @@ def __init__(
...
-        self._quad = get_quadrature(setdefault(quad, default_quad), automorphism)
+        self._quad = get_quadrature(
+            setdefault(
+                quad,
+                get_quadrature(
+                    leggauss(32),
+                    (automorphism_sin, grad_automorphism_sin),
+                ),
+            ),
+            automorphism,
```
Comment (unalmis) [inline desc/integrals/bounce_integral.py]: - This will now solve an eigenvalue problem every time `Bounce*D` is called. 
- Autodiff will also trace it into the graph as well now, and JIT is imperfect and does not work well inside scan loops, which is where this stuff is called.
For error analysis, it is vital that the quadrature is not differentiated, and leaving that up to JIT under scan is not something I want given how often bugs are found in it. thus, this change makes it harder to confirm correctness now.
- On the other hand, what I suggested in the original issue of just changing the default_quad defined in this file globally avoids an eigenvalue solve on import.

File: desc/integrals/bounce_integral.py [hunk 2]
```diff
@@ -1468,7 +1471,13 @@ def __init__(
     ):
         """Returns an object to compute bounce integrals."""
         assert grid.is_meshgrid
-        quad = setdefault(quad, default_quad)
+        quad = setdefault(
+            quad,
+            get_quadrature(
+                leggauss(32),
+                (automorphism_sin, grad_automorphism_sin),
+            ),
```
Comment (unalmis) [inline desc/integrals/bounce_integral.py]: https://github.com/PlasmaControl/DESC/pull/2156/changes#r3062003714

File: desc/backend.py [hunk 3]
```diff
@@ -36,13 +36,11 @@
                     + "installed JAX with GPU support?"
                 )
                 set_device("cpu")
-            x = jnp.linspace(0, 5)
-            y = jnp.exp(x)
+            x = jnp.arange(2)
```
Comment (unalmis) [inline desc/backend.py]: i don't know what the purpose of this change is, but generally you want to do math to test this stuff is working. arange and the like can often execute on host on old jax (< 0.9) versions

**Review by unalmis — DISMISSED** [review]
my suggestion on the issue avoids an eigenvalue solve on import. current changes force an eigenvalue solve every time an object is made, which is worse for optimization and debugging with jit in general. i didn't check the claude stuff

File: desc/integrals/bounce_integral.py (hunk 1 above)
Comment (YigitElma) [reply desc/integrals/bounce_integral.py]: https://github.com/PlasmaControl/DESC/issues/2154#issuecomment-4216548470
This was my way of making the default_quad required(ish) argument; if someone wants to avoid this computation, they can pass a precomputed value. If you want to make it actually required, then you can make the change. My point is, there shouldn't be any computation on the import; not everyone uses this function.

File: desc/backend.py (hunk 3 above)
Comment (YigitElma) [reply desc/backend.py]: I think to debug any problem with `jax`, any computation that triggers XLA is enough, at least until now I've never seen otherwise. Pure imports don't show actual problems, but any `jnp` function does the job. I just made the change to get that check minimally. I don't have a strong opinion about it, but I prefer something that is minimal.

File: desc/integrals/bounce_integral.py (hunk 1 above)
Comment (unalmis) [reply desc/integrals/bounce_integral.py]: >  if someone wants to avoid this computation, they can pass a precomputed value.

This is false. Your changes force an eigenvalue solve every time Bounce*D is called

Comment (unalmis) [reply desc/integrals/bounce_integral.py]: - Trivial computations like `jnp.cos(jnp.arange(32))` on import are done by the packages we import like numpy etc. For example they load up things like powers of two into cache just to name one. unlike avoiding an eigenvalue solve, avoiding calling a cosine function is probably an unnecessary micro-optimization.
- jnp.cos(jnp.arange(32)) probably consumes comparable time to defining a python function in some file. So it doesn't make sense to me to single that out as a computation. everything is machine code; everything is a computation.

- Previously, users could import `default_quad`, a public variable, into their scripts. They can no longer do that with your change. User code will thus break.
- very simple solution is to just use a the default quad that avoids eigenvalue solve as mentioned in original issue.

Comment (f0uriest) [reply desc/integrals/bounce_integral.py]: The main cost of `leggauss(32)` is the jit compiler overhead. The actual calculation takes ~1ms. So one option (if we want to keep using legendre for default quad) could be to just use the original `np.polynomial.legendre.leggauss` for `default_quad`, which avoids the jit overhead at init (and since its just a constant it doesn't really need to be jax anyways). switching to chebyshev quadrature avoids the eigenvalue solve but may still have some jit overhead leading to similar issues. (although in either case a pure np version may be faster). I'll defer to kaya on the accuracy impact of legendre vs chebyshev

File: desc/backend.py (hunk 3 above)
Comment (f0uriest) [reply desc/backend.py]: I'd prefer to keep the old version. As kaya mentioned, older versions of jax will always execute `arange` on the host meaning GPU issues won't necessarily show up.

File: desc/compute/__init__.py [hunk 4]
```diff
@@ -55,42 +54,160 @@
...
+        return transforms, params, profiles
 
     for p in data_index:
-        for key in data_index[p]:
+        # --- Step 1: Topological sort via iterative Depth-First Search ---
+        # We need to process quantities with no dependencies before the
+        # quantities that depend on them. This way, when we process key K,
+        # all of K's dependencies already have their full_dependencies and
+        # full_with_axis_dependencies cached, and we can build K's full
+        # dependency set with a simple set union instead of deep recursion.
```
Comment (f0uriest) [inline desc/compute/__init__.py:110]: This sounds like it's basically doing what `set_tier` is doing below, so ideally we could do them both at the same time.

Comment (YigitElma) [reply desc/compute/__init__.py:110]: Yeah, I was actually gonna suggest that in looped compute PR. The order obtained at the end can be used directly. That is why I have special sort instead of normal sort.

File: desc/integrals/bounce_integral.py (hunk 1 above)
Comment (unalmis) [reply desc/integrals/bounce_integral.py]: accuracy is negligible. does jnp.asarray(numpy computation) have jit overhead? don't want issues like this where jax bakes in the numpy constant: https://github.com/PlasmaControl/DESC/blob/2d9d35a29fa9c1e3e9158ced059ffc8932fa3970/desc/integrals/surface_integral.py#L285

Comment (ddudt) [reply desc/integrals/bounce_integral.py]: The only place that `default_quad` is ever used is to set the default value in `Bounce1D` and `Bounce2D` if `quad=None`. This rarely happens in practice, since the compute functions always pass a value for `quad` when creating instances of these classes. 

My preference is to remove it from the import overhead and create it in the `Bounce*D.__init__` methods only when needed. Something like this:
```
if quad is None:
    quad = leggauss(32)
if automorphism is None:
    automorphism = (automorphism_sin, grad_automorphism_sin)
self._quad = get_quadrature(quad, automorphism)
```
Or document those as the default values and then we only need:
```
self._quad = get_quadrature(quad, automorphism)
```
That should only add the extra compute in the rare instances when the user doesn't specify `quad` themselves. 

I'm also fine with switching to `chebgauss1` or the numpy version if the accuracy is negligible. But I still think it should happen inside the `Bounce*D` classes instead of on import, since it mostly likely never gets used. I'm not worried about removing `default_quad` from the "public API" since it is not publicly documented anyways.

Comment (unalmis) [reply desc/integrals/bounce_integral.py]: Daniel code will break convergence by applying automorphism to all input quadrature.

I will be fine with this:
```python
if quad is None:
    quad = jax.lax.stop_gradient(get_quadrature(leggauss(32), (automorphism_sin, grad_automorphism_sin)))
elif automorphism is not None:
    quad = get_quadrature(quad, automorphism))
self._quad = quad # or put stop grad here
```

Comment (unalmis) [reply desc/integrals/bounce_integral.py]: yea its probably better to always call stop grad to also prevent loss of numerical precision since differentiating the quadrature makes these integrals non-integrable, and we don't want to rely on floating point math to cancel non-integrable singularity

```python
if quad is None:
    quad = get_quadrature(leggauss(32), (automorphism_sin, grad_automorphism_sin))
elif automorphism is not None:
    quad = get_quadrature(quad, automorphism)
self._quad = jax.lax.stop_gradient(quad)
```

we had to call stop grad to avoid floating point issues in stability stuff too

Comment (unalmis) [reply desc/integrals/bounce_integral.py]: leggaus works on nonsingular stuff too so its more dummy proof for users so keep that

File: desc/integrals/bounce_integral.py [hunk 5]
```diff
@@ -301,7 +295,14 @@ def __init__(
         is_reshaped = is_reshaped or is_fourier
         vander = setdefault(vander, {})
 
-        self._quad = get_quadrature(setdefault(quad, default_quad), automorphism)
+        if quad is None:
+            quad = get_quadrature(
+                leggauss(32), (automorphism_sin, grad_automorphism_sin)
+            )
+        elif automorphism is not None:
```
Comment (unalmis) [inline desc/integrals/bounce_integral.py]:
```suggestion
        else:
````

File: desc/integrals/bounce_integral.py [hunk 6]
```diff
@@ -1468,9 +1469,13 @@ def __init__(
...
         """Returns an object to compute bounce integrals."""
         assert grid.is_meshgrid
-        quad = setdefault(quad, default_quad)
-
-        self._quad = get_quadrature(quad, automorphism)
+        if quad is None:
+            quad = get_quadrature(
+                leggauss(32), (automorphism_sin, grad_automorphism_sin)
+            )
+        elif automorphism is not None:
```
Comment (unalmis) [inline desc/integrals/bounce_integral.py]:
```suggestion
        else:
````

**Review by unalmis — APPROVED** [review]

**Comment by unalmis** [comment]
ididn't check calude stuff

File: .gitignore [hunk 7]
```diff
@@ -30,3 +30,5 @@ uv.lock
 # Environments
 .env
 .venv
+
```
Comment (ddudt) [inline .gitignore:33]: accidental line break?

**Review by ddudt — APPROVED** [review]

File: .gitignore (hunk 7 above)
Comment (YigitElma) [reply .gitignore:33]: I didn't want to put it directly under the environment tab

File: desc/compute/__init__.py (hunk 4 above)
Comment (YigitElma) [reply desc/compute/__init__.py:110]: https://github.com/PlasmaControl/DESC/pull/1557#discussion_r3068395627

**Review by dpanici — APPROVED** [review]

**Comment by YigitElma** [comment]
I will merge this after the dev meeting, in case @f0uriest has some additional comments

**Comment by YigitElma** [comment]
As a final check, I used this. Save both `data_index` objects in master and this PR (I have to remove the `fun` since it cannot be pickled.
```python
import pickle

data_index_new = data_index.copy()

for p in data_index_new:
    for key in data_index_new[p]:
        data_index_new[p][key].pop("fun")
output = open("data_index_pr.pkl", "wb")

# Pickle dictionary using protocol 0.
pickle.dump(data_index, output)
output.close()
```
Then compare them.
```python
pkl1 = open("data_index_master.pkl", "rb")
pkl2 = open("data_index_pr.pkl", "rb")

data1 = pickle.load(pkl1)
data2 = pickle.load(pkl2)

def norm(full):
    # sort alphabetically (since PR sorts in topo order)
    return {
        "data": sorted(full["data"]),
        "params": sorted(full["params"]),
        "profiles": sorted(full["profiles"]),
        "transforms": {k: sorted(v) for k, v in full["transforms"].items()},
    }

for p in data1:
    if p not in data2:
        print(f"{p} not in pr data_index")
        continue
    for k in data1[p]:
        if k not in data2[p]:
            print(f"{p}-{k} not in pr data_index")
            continue
        fd1 = norm(data1[p][k]["full_dependencies"])
        fd2 = norm(data2[p][k]["full_dependencies"])
        assert fd1 == fd2, f"full_dependencies differ for {p}-{k}"
        fa1 = norm(data1[p][k]["full_with_axis_dependencies"])
        fa2 = norm(data2[p][k]["full_with_axis_dependencies"])
        assert fa1 == fa2, f"full_with_axis_dependencies differ for {p}-{k}"

for p in data2:
    if p not in data1:
        print(f"{p} not in master data_index")
        continue
    for k in data2[p]:
        if k not in data1[p]:
            print(f"{p}-{k} not in master data_index")
            continue
        fd1 = norm(data1[p][k]["full_dependencies"])
        fd2 = norm(data2[p][k]["full_dependencies"])
        assert fd1 == fd2, f"full_dependencies differ for {p}-{k}"
        fa1 = norm(data1[p][k]["full_with_axis_dependencies"])
        fa2 = norm(data2[p][k]["full_with_axis_dependencies"])
        assert fa1 == fa2, f"full_with_axis_dependencies differ for {p}-{k}"

pkl1.close()
pkl2.close()
```

And this passes.

**Comment by YigitElma** [comment]
> I will merge this after the dev meeting, in case @f0uriest has some additional comments

Rory said he is fine, so I am merging now.
